java/openmldb-spark-connector/src/main/java/com/_4paradigm/openmldb/spark/write/OpenmldbDataWriter.java

```diff
@@ -0,0 +1,154 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com._4paradigm.openmldb.spark.write;
+
+import com._4paradigm.openmldb.sdk.Schema;
+import com._4paradigm.openmldb.sdk.SdkOption;
+import com._4paradigm.openmldb.sdk.SqlException;
+import com._4paradigm.openmldb.sdk.impl.SqlClusterExecutor;
+import com.google.common.base.Preconditions;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.connector.write.DataWriter;
+import org.apache.spark.sql.connector.write.WriterCommitMessage;
+
+import java.io.IOException;
+import java.sql.Date;
+import java.sql.PreparedStatement;
+import java.sql.ResultSetMetaData;
+import java.sql.SQLException;
+import java.sql.Timestamp;
+import java.sql.Types;
+
+public class OpenmldbDataWriter implements DataWriter<InternalRow> {
+    private final int partitionId;
+    private final long taskId;
+    private PreparedStatement preparedStatement = null;
+
+    public OpenmldbDataWriter(OpenmldbWriteConfig config, int partitionId, long taskId) {
+        try {
+            SdkOption option = new SdkOption();
+            option.setZkCluster(config.zkCluster);
+            option.setZkPath(config.zkPath);
+            SqlClusterExecutor executor = new SqlClusterExecutor(option);
+            String dbName = config.dbName;
+            String tableName = config.tableName;
+
+            Schema schema = executor.getTableSchema(dbName, tableName);
+            // create insert placeholder
+            StringBuilder insert = new StringBuilder("insert into " + tableName + " values(?");
+            for (int i = 1; i < schema.getColumnList().size(); i++) {
+                insert.append(",?");
+            }
+            insert.append(");");
+            preparedStatement = executor.getInsertPreparedStmt(dbName, insert.toString());
+        } catch (SQLException | SqlException e) {
+            e.printStackTrace();
+        }
+
+        this.partitionId = partitionId;
+        this.taskId = taskId;
+    }
+
+    @Override
+    public void write(InternalRow record) throws IOException {
+        try {
+            // record to openmldb row
+            ResultSetMetaData metaData = preparedStatement.getMetaData();
+            Preconditions.checkState(record.numFields() == metaData.getColumnCount());
+            addRow(record, preparedStatement);
+            preparedStatement.addBatch();
+        } catch (SQLException e) {
+            e.printStackTrace();
+            throw new IOException(e.getMessage());
+        }
+    }
+
+    private void addRow(InternalRow record, PreparedStatement preparedStatement) throws SQLException {
+        // idx in preparedStatement starts from 1
+        ResultSetMetaData metaData = preparedStatement.getMetaData();
+        for (int i = 0; i < record.numFields(); i++) {
```

Comment:
We may start with 1 for the for-loop.

Reply:
but in record, fields start from 0, actually it's a vector of columns.

Reply:
Oh, did not see the usage of `record`.